python/lsst/meas/algorithms/stamps.py

```diff
@@ -144,11 +150,26 @@ def readFitsWithOptions(filename, stamp_factory, options):
                 bbox = Box2I(Point2I(llcX, llcY), Extent2I(width, height))
                 kwargs["bbox"] = bbox
         stamp_parts = {}
+
+        # Determine the dtype from the factory. This allows a Stamp class
+        # to be defined in terms of MaskedImageD or MaskedImageI without
+        # forcing everything to floats.
+        masked_image_cls = None
+        for stamp_field in fields(stamp_factory.__self__):
+            if issubclass(stamp_field.type, MaskedImage):
+                masked_image_cls = stamp_field.type
+                break
+        else:
+            raise RuntimeError("Stamp factory does not use MaskedImage.")
+        default_dtype = np.dtype(masked_image_cls.dtype)
+
         # We need to be careful because nExtensions includes the primary HDU.
         for idx in range(nExtensions - 1):
+            dtype = None
             md = readMetadata(filename, hdu=idx + 1)
             if md["EXTNAME"] in ("IMAGE", "VARIANCE"):
                 reader = ImageFitsReader(filename, hdu=idx + 1)
+                dtype = default_dtype
```

Comment:
Variances are never int, I think.

Reply:
Thanks. Yes, MaskedImageD and MaskedImageI both use ImageF. I find it a bit confusing that a double precision image should come with a single precision variance but that's what is happening so I'll force dtype to None for VARIANCE as well as MASK.